accepted/2021/runtime-security-mitigations.md

```diff
@@ -0,0 +1,213 @@
+- [Appendix](#appendix)
+  - [Timing Side Channel Attacks](#timing-side-channel-attacks)
+  - [Confidential Computing](#confidential-computing)
+
+# Anatomy of an exploit
+
+Memory security vulnerability starts with a memory safety violation. That includes bugs such as writing outside of an array's bounds or using a pointer after freeing the associated object. When an attacker can influence the pointer or data involved, they have at least some control of the program's memory (this is referred to as a write-what-where primitive). Once an attacker has a write-what-where primitive, they may be able to use that to change the program's execution to do something malicious.
+
+The security mitigations prevent or reduce probability of each step in this sequence.
+
+The mitigations for memory safety violations must be typically applied to all code running in a process to be fully effective. The exploits often combine memory safety violations and code from multiple modules in the process. For example, memory safety violation in C/C++ module and code generated by .NET runtime JIT can be combined together to produce an exploit. Thus, missing mitigations in .NET modules can allow exploiting vulnerabilities in C/C++ modules that would be hard or impossible to exploit in isolation, and vice versa.
+
+# Executable Space Protection
+
+_Known as Data Execution Prevention (DEP) on Windows_
+
+The simplest path for an attack is to write new malicious code in memory and then get control flow to jump to it. Executable space protection enforces that memory has independent write and execution permissions. Attacks can no longer create new code because executable memory is not writable and writable memory is not executable.
+
+## W^X
+
+[W^X](https://en.wikipedia.org/wiki/W%5EX) is one of the most fundamental mitigations. It blocks the simplest attack path by disallowing memory pages to be writeable and executable at the same time. .NET runtime has been missing this mitigation so far and the lack of this mitigation has (correctly) resulted in us not considering more advanced mitigations. Large number of pages that are both writeable and executable in a typical .NET process is a ripe target for attacks that simply inject new code.
+
+Apple has made the W^X mandatory for future versions of macOS desktop operating system as part of Apple Silicon transition. It motivated us to schedule implementation of this mitigation for .NET 6, on all supported operating systems. Our principle is to treat all supported operating systems equally with respect to security, where possible.
+
+## Protect Intermediate Language (IL) Code
+
+IL code is equivalent of executable code. The IL code might be overwritten with malicious IL. An attacker could then wait for the IL to be compiled or re-compiled via tiering and have malicious code injected that way.
+
+.NET runtime should treat the IL code same way as directly executable code, ie make IL read-only or otherwise protected wherever possible.
```

Comment:
Also seems like an area where we should avoid applying certain kinds of optimizations.  For example, DynamicMethod fits squarely in the target area of writeable managed memory for the purposes of generating code to be executed.  Today, it doesn't use ArrayPool, but there have been suggestions in the past that it do so for some of the potentially large arrays it creates.  As part of reducing avenues of attack, this is a case where we should probably avoid using pooled arrays (we probably wouldn't have done so anyway due to poor tradeoffs, but wanted to throw it out as an example).